### PR TITLE
Resolves Search Issues #1383 #1384 #1388

### PR DESCRIPTION
**Pull request by prb112**
## refactor: Values Table have duplicate data for _id and _lastUpdated #1388
- FHIRPersistenceJDBCImpl filters the codes in the
extractParameterValues step and keeps the parameter values from hitting
the corresponding tables

## fix: Chained Search with _id result in Http Status 500 #1384

URL
`https://localhost:9443/fhir-server/api/v4/Procedure?_count=10&subject:Patient._id=173a1657ba3-b9399454-1c17-4b96-9665-6cb930bc9c36&_page=1`

GENERATES SQL

`SELECT R.RESOURCE_ID, R.LOGICAL_RESOURCE_ID, R.VERSION_ID, R.LAST_UPDATED, R.IS_DELETED, R.DATA, LR.LOGICAL_ID  FROM PROCEDURE_RESOURCES R JOIN (SELECT DISTINCT LR.LOGICAL_RESOURCE_ID, LR.LOGICAL_ID, LR.CURRENT_RESOURCE_ID FROM Procedure_LOGICAL_RESOURCES LR  JOIN Procedure_RESOURCES R ON R.LOGICAL_RESOURCE_ID = LR.LOGICAL_RESOURCE_ID   AND R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID   AND R.IS_DELETED <> 'Y'  JOIN Procedure_STR_VALUES  AS param0 ON (param0.PARAMETER_NAME_ID=1396 AND (param0.STR_VALUE IN (SELECT 'Patient' || '/' || CLR1.LOGICAL_ID FROM Patient_RESOURCES CR1, Patient_LOGICAL_RESOURCES CLR1, Patient_TOKEN_VALUES CP1 WHERE CR1.RESOURCE_ID = CLR1.CURRENT_RESOURCE_ID AND CR1.IS_DELETED <> 'Y' AND CP1.LOGICAL_RESOURCE_ID = CR1.LOGICAL_RESOURCE_ID AND CLR1.LOGICAL_ID IN (?) ))) AND LR.LOGICAL_RESOURCE_ID = param0.LOGICAL_RESOURCE_ID) AS LR  ON      R.LOGICAL_RESOURCE_ID = LR.LOGICAL_RESOURCE_ID  AND R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID  AND R.IS_DELETED <> 'Y' ORDER BY RESOURCE_ID ASC  OFFSET 0 ROWS FETCH NEXT 10 ROWS ONLY  searchArgs=[173a1657ba3-b9399454-1c17-4b96-9665-6cb930bc9c36] executionTime=26.225081ms`

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/QuerySegmentAggregator.java [hunk 1]
```diff
@@ -196,15 +196,16 @@ protected SqlQueryData buildQuery() throws Exception {
             queryString.append("     R.LOGICAL_RESOURCE_ID = LR.LOGICAL_RESOURCE_ID ");
             queryString.append(" AND R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID ");
             queryString.append(" AND R.IS_DELETED <> 'Y'");
-            
 
-            // Bind Variables
+
+            // An important step here is to add _id, values table bind variables, and then and _lastUpdated
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/QuerySegmentAggregator.java:201]:
```suggestion
            // An important step here is to add _id, values table bind variables, and then _lastUpdated
```

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/QuerySegmentAggregator.java [hunk 2]
```diff
@@ -267,13 +268,14 @@ protected SqlQueryData buildCountQuery() throws Exception {
             buildFromClause(queryString, simpleName);
             buildWhereClause(queryString, null);
 
-            // An important step here is to add _id then all other values. 
+            // An important step here is to add _id, values table bind variables, and then and _lastUpdated
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/QuerySegmentAggregator.java:271]:
```suggestion
            // An important step here is to add _id, values table bind variables, and then _lastUpdated
```

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/QuerySegmentAggregator.java [hunk 3]
```diff
@@ -358,20 +359,19 @@ protected SqlQueryData buildSystemLevelQuery(String selectRoot, String subSelect
...
-                allBindVariables.addAll(idsObjects);
-                allBindVariables.addAll(lastUpdatedObjects);
 
                 queryString.append(ON);
                 queryString.append("     R.LOGICAL_RESOURCE_ID = LR.LOGICAL_RESOURCE_ID ");
                 queryString.append(" AND R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID ");
                 queryString.append(" AND R.IS_DELETED <> 'Y'");
-                
+
+                // An important step here is to add _id, values table bind variables, and then and _lastUpdated
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/QuerySegmentAggregator.java:368]:
```suggestion
                // An important step here is to add _id, values table bind variables, and then _lastUpdated
```

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/QuerySegmentAggregator.java [hunk 4]
```diff
@@ -604,7 +606,10 @@ protected void buildWhereClause(StringBuilder whereClause, String overrideType)
                         .append(".LOGICAL_RESOURCE_ID = R.LOGICAL_RESOURCE_ID");
                     }
                 }
-            } // end if SKIP_WHERE
+            } else {
+                // end if SKIP_WHERE
+                System.out.println(whereClause);
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/QuerySegmentAggregator.java:611]: did you mean to take this back out?  if not, it should be a log statement, not sysout

Comment (prb112) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/QuerySegmentAggregator.java:611]: Yes, I meant to remove.  Fixed in next commit.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/QuerySegmentAggregator.java (hunk 3 above)
Comment (prb112) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/QuerySegmentAggregator.java:366]: Fixed

**Comment by prb112** [comment]
Right. I'll fix that part

**Comment by prb112** [comment]
> The join
> 
> ```
> SELECT 'Patient' || '/' || CLR1.LOGICAL_ID
>                                        FROM Patient_RESOURCES CR1,
>                                             Patient_LOGICAL_RESOURCES CLR1,
>                                             Patient_TOKEN_VALUES CP1
>                                       WHERE CR1.RESOURCE_ID = CLR1.CURRENT_RESOURCE_ID
>                                         AND CR1.IS_DELETED <> 'Y'
>                                         AND CP1.LOGICAL_RESOURCE_ID = CR1.LOGICAL_RESOURCE_ID
>                                         AND CLR1.LOGICAL_ID IN (?) 
> ```
> 
> is incorrect. There's no need to join patient_token_values in this case, and doing so without any filter predicate will be very expensive.

```
SELECT R.RESOURCE_ID, R.LOGICAL_RESOURCE_ID, R.VERSION_ID, R.LAST_UPDATED, R.IS_DELETED, R.DATA, LR.LOGICAL_ID
FROM PROCEDURE_RESOURCES R JOIN (
    
    
    
    SELECT DISTINCT LR.LOGICAL_RESOURCE_ID, LR.LOGICAL_ID, LR.CURRENT_RESOURCE_ID
    FROM Procedure_LOGICAL_RESOURCES LR JOIN Procedure_RESOURCES R ON R.LOGICAL_RESOURCE_ID = LR.LOGICAL_RESOURCE_ID AND R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID AND R.IS_DELETED <> 'Y' JOIN Procedure_STR_VALUES  AS param0 ON (param0.PARAMETER_NAME_ID=1396 AND (param0.STR_VALUE IN (
        
        
        
        SELECT 'Patient' || '/' || CLR1.LOGICAL_ID
            FROM Patient_RESOURCES CR1, Patient_LOGICAL_RESOURCES CLR1
            WHERE CR1.RESOURCE_ID = CLR1.CURRENT_RESOURCE_ID AND CR1.IS_DELETED <> 'Y' AND CLR1.LOGICAL_ID IN (?) ))) AND LR.LOGICAL_RESOURCE_ID = param0.LOGICAL_RESOURCE_ID
        
        
        
        
        ) AS LR ON      R.LOGICAL_RESOURCE_ID = LR.LOGICAL_RESOURCE_ID AND R.RESOURCE_ID = LR.CURRENT_RESOURCE_ID AND R.IS_DELETED <> 'Y'
ORDER BY RESOURCE_ID ASC  LIMIT 10 OFFSET
0 /* <OPTGUIDELINES> <REOPT VALUE='ONCE'/> </OPTGUIDELINES> */ 
```

This is the full query, I spaced it out to give context to the snippet I had before.

**Comment by prb112** [comment]
I've made the change in the next commit

**Comment by punktilious** [comment]
Execution plan looks good.

**Pull request merged**